File: local/migration/report.py
"""Migration report entries and aggregation."""

from __future__ import annotations

from enum import Enum
from typing import Any

from pydantic import BaseModel, Field


class Severity(str, Enum):
    AUTO = "auto"
    APPROXIMATION = "approximation"
    MANUAL_REQUIRED = "manual_required"
    BLOCKER = "blocker"


class ReportEntry(BaseModel):
    severity: Severity
    category: str
    message: str
    source_line: str | None = None
    pan_hint: str | None = None


class MigrationReport(BaseModel):
    source_format: str = "unknown"
    entries: list[ReportEntry] = Field(default_factory=list)
    unmapped_lines: list[str] = Field(default_factory=list)

    def add(
        self,
        severity: Severity,
        category: str,
        message: str,
        *,
        source_line: str | None = None,
        pan_hint: str | None = None,
    ) -> None:
        self.entries.append(
            ReportEntry(
                severity=severity,
                category=category,
                message=message,
                source_line=source_line,
                pan_hint=pan_hint,
            )
        )

    def summary(self) -> dict[str, int]:
        counts: dict[str, int] = {}
        for e in self.entries:
            counts[e.severity.value] = counts.get(e.severity.value, 0) + 1
        counts["unmapped_lines"] = len(self.unmapped_lines)
        return counts

    def to_dict(self) -> dict[str, Any]:
        return {
            "source_format": self.source_format,
            "summary": self.summary(),
            "entries": [e.model_dump() for e in self.entries],
            "unmapped_lines": self.unmapped_lines,
        }

File: dotnet/Services/Migration/Report.cs
using System.Text.Json.Nodes;

namespace PanCopilot.Services.Migration;

// Port of migration/report.py.
public enum Severity { Auto, Approximation, ManualRequired, Blocker }

public static class SeverityExt
{
    public static string Value(this Severity s) => s switch
    {
        Severity.Auto => "auto",
        Severity.Approximation => "approximation",
        Severity.ManualRequired => "manual_required",
        Severity.Blocker => "blocker",
        _ => "auto",
    };
}

public sealed class ReportEntry
{
    public Severity Severity { get; init; }
    public string Category { get; init; } = "";
    public string Message { get; init; } = "";
    public string? SourceLine { get; init; }
    public string? PanHint { get; init; }
}

public sealed class MigrationReport
{
    public string SourceFormat { get; set; } = "unknown";
    public List<ReportEntry> Entries { get; } = new();
    public List<string> UnmappedLines { get; } = new();

    public void Add(Severity severity, string category, string message,
        string? sourceLine = null, string? panHint = null) =>
        Entries.Add(new ReportEntry
        {
            Severity = severity, Category = category, Message = message,
            SourceLine = sourceLine, PanHint = panHint,
        });

    public Dictionary<string, int> Summary()
    {
        var counts = new Dictionary<string, int>();
        foreach (var e in Entries)
        {
            var k = e.Severity.Value();
            counts[k] = counts.GetValueOrDefault(k) + 1;
        }
        counts["unmapped_lines"] = UnmappedLines.Count;
        return counts;
    }

    public JsonObject ToJson()
    {
        var entries = new JsonArray();
        foreach (var e in Entries)
            entries.Add(new JsonObject
            {
                ["severity"] = e.Severity.Value(),
                ["category"] = e.Category,
                ["message"] = e.Message,
                ["source_line"] = e.SourceLine,
                ["pan_hint"] = e.PanHint,
            });
        var summary = new JsonObject();
        foreach (var kv in Summary()) summary[kv.Key] = kv.Value;
        var unmapped = new JsonArray();
        foreach (var l in UnmappedLines) unmapped.Add(l);
        return new JsonObject
        {
            ["source_format"] = SourceFormat,
            ["summary"] = summary,
            ["entries"] = entries,
            ["unmapped_lines"] = unmapped,
        };
    }
}
